### PR TITLE
Networking See PR #532

### PR DESCRIPTION
**Pull request by jojolepro**
See PR #532 for all comments. This is just the same re-opened because bors won't merge the other one.

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/665)

### CODE REVIEW

**Comment by Xaeroxe** [comment]
cache got too big 0_o

bors retry

**Comment by jojolepro** [comment]
nice

**Comment by Xaeroxe** [comment]
bors retry

**Comment by jojolepro** [comment]
>Designated thread does polling mio and sends tasks to thread-pool.

We'll have thread starvation issues, as we have with audio already when the game is under heavy load.

>Network events and input events can be handled by that system together as there are similarities. Like character controlling events from players local and remote.

You can do that by simply using a EventChannel<NetEvent<T>> in any system you want.

>And I'd want real examples. Like sending pong state over the network.

Me too, but I have to wait for specs 0.11. I also have to rethink how I want the entity ownership to work. Mainly, I need to tag components I want to sync (Marked?) and then find a way to deserialize them on the receiving end without actually having the type. Basically the same problem that I had with NetEvent, but I can't use an enum...

>Also it may be better to use more abstract data sources than sockets.

I already talked about it. Recap: Its a planned feature, but it will require a completely separate abstraction PR for it. (I'm planning on doing it in the 3rd one, while this is the first one. The second one will be component/entity sync/creation/destruction/ownerships).

>There are crates for async IO which should be useful here.

Except for tokio and a couple of lower level subcrates, I didn't see much that were promising and didn't require a completely separate thread. Since most people are against using futures, the choice is pretty limited.

File: amethyst_network/src/net_event.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        /// The reason of the disconnection.
+        reason: String,
+    },
+    /// Notify the clients(including the one being disconnected) that a client has been disconnected from the server.
+    Disconnected {
+        /// The reason of the disconnection.
+        reason: String,
+    },
+    /// A simple text message event.
+    TextMessage {
```
[the remark this answers is not in the document]
Comment (jojolepro) [reply amethyst_network/src/net_event.rs]: info!("")
Those are arbitrary debugging messages.
For example "Client 1. Time is: 3254, ping is: 6451, client state: Connected".

File: amethyst_network/src/net_event.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    },
+    /// Tell the server that the client is disconnecting.
+    Disconnect {
+        /// The reason of the disconnection.
+        reason: String,
+    },
+    /// Notify the clients(including the one being disconnected) that a client has been disconnected from the server.
+    Disconnected {
+        /// The reason of the disconnection.
+        reason: String,
```
[the remark this answers is not in the document]
Comment (jojolepro) [reply amethyst_network/src/net_event.rs]: true

**Comment by jojolepro** [comment]
Thanks for the review, btw ;)

File: amethyst_network/src/net_event.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (omni-viral) [reply amethyst_network/src/net_event.rs]: I can't see it from variant name though

[the remark this answers is not in the document]
Comment (jojolepro) [reply amethyst_network/src/net_event.rs]: I can rename it DebugTextMessage.

**Comment by omni-viral** [comment]
> We'll have thread starvation issues

I see network polling once in 17ms (60 fps) as starved already. Also. If we will have control over that thread we could ensure it does polling frequently enough. With audio we have no control over playback thread.

> You can do that by simply using a EventChannel<NetEvent>

Is `EventChannel<NetEvent>` capable to broadcast network events under heavy load?

> Since most people are against using futures

I can't understand why people are against futures. We tried to use them wrong and went into troubles. Well. Next time we can try to use them right 😏

**Comment by jojolepro** [comment]
>Is EventChannel<NetEvent> capable to broadcast all of network events under heavy load?

Yes. I made it run at 10k events per frame.

>I can't understand why people are against futures. We tried to use them wrong and went into troubles. Well. Next time we can try to use them right

I like futures, but the problem is that they can complete at any time. Forking the execution thread is easy but merging it back into the main execution loops is much harder. The way to solve it is to use the equivalent of unity's coroutines: You run your future normally, but when it completes, it stores the result in some temporary buffer (locking). Then from the main loop (in the sense that it has to happen during an update() frame) you get that result and call the action that is supposed to happen. This is a limitation inherent to the fact that amethyst is frame-based.

Frame based: It starts with a state, applies a set of instructions on it and ends up with a new step.

Futures are better with completely async systems: You can fork the execution as much as you want because you are not required to have a global end state to compute.

I didn't mean to write so much on this lol

**Comment by jojolepro** [comment]
Its probably possible to have a separate thread to poll the network socket, but the events will be consumed by the ecs only when the update() loop is there. Might as well just poll the socket during the loop at that point. Mio prevents packet loss slow stepping, from what I have seen.

**Comment by jojolepro** [comment]
Its a separate topic, but if someone can come up with an architecture that allow using Futures in a stepping engine without requiring locking buffers (or buffers at all), please open a rfc.

**Comment by jojolepro** [comment]
Currently refactoring, the solution just jumped in my face. Instead of having things like NetConnectionPool that is a resource containing all the connections, then try to bind that to a NetIdentity or whatever the heck I was doing, I'll just create an entity and put all the things the connection OWNS on the entity. Not sure why I'm so blind that I didn't see that before.

**Comment by LucioFranco** [comment]
Closing this PR as it still needs some work, that @jojolepro and I have talked about, and since we are moving to master #924 this will need to be reopened against that branch.

**Pull request closed without merging**